connectors/okta/connector/data_handler.py

```diff
@@ -0,0 +1,255 @@
+
+    def create_source_report_object(self):
+        if not (self.source and self.report):
+            # create source and report entry
+            self.source = {'_key': context().args.source,
+                           'name': "Okta",
+                           'description': 'Identity and Access Management service',
+                           'product_link': "https://www.okta.com"}
+            self.report = {'_key': str(self.timestamp),
+                           'timestamp': self.timestamp,
+                           'type': 'Okta',
+                           'description': 'Okta reports'}
+        return {'source': self.source, 'report': self.report}
+
+    # Create asset Object as per CAR data model from data source
+    def handle_asset(self, obj):
+        """create asset object"""
+        if obj:
+            # user login client as asset
+            for asset in obj:
+                res = {}
+                if asset.get('client'):
+                    user_id = asset['actor']['id']
+                    # handling users assigned to an application
+                    # this case asset, account nodes are present,
+                    # we are adding asset_account edge.
+                    if asset['eventType'] == 'application.user_membership.add':
+                        for record in asset['target']:
+                            if record['type'] == 'AppInstance':
+                                res['external_id'] = record['id']
+                            if record['type'] == 'User':
+                                user_id = record['id']
+                    else:
+                        browser = asset['client']['userAgent']['browser']
+                        os = asset['client']['userAgent']['os']
+                        ipaddress = asset['client']['ipAddress']
+                        device = asset['client']['device']
+                        res['name'] = f'{browser} on {os} {device} from {ipaddress}'
+                        res['external_id'] = get_asset_id(asset)
+                        self.add_collection('asset', res, 'external_id')
+                        # asset_account edge
+                    self.add_edge('asset_account', {'_from_external_id': res['external_id'],
+                                                    '_to_external_id': user_id})
+
+                else:
+                    res['external_id'] = asset['id']
+                    res['name'] = asset['label']
+                    self.add_collection('asset', res, 'external_id')
+
+                    # Handle asset_account
+                    for user in asset.get('users'):
+                        self.add_edge('asset_account', {'_from_external_id': res['external_id'],
+                                                        '_to_external_id': user['id']})
+
+    # Create user Object as per CAR data model from data source
+    def handle_user(self, obj):
+        """create user object"""
+        if obj:
+            for asset in obj:
```

Comment:
Do you mean `user` instead of `asset` here?

Reply:
Yes, your understanding is correct.
For better understanding user and account handler functions identifiers are updated